### PR TITLE
Feature/premakedeps

### LINKED ISSUE #13275: [bug] Error in generator 'PremakeDeps': '_Component' object has no attribute 'aggregate_components'
### Environment details

* Operating System+version: Windows 11 (10.0.22621 Build 22621)
* Compiler+version: MSVC 193
* Conan version: 2.0.0
* Python version: 3.11.0

### Steps to reproduce

1. Create an empty/default conan file (Tried both .txt and .py --> same result).
2. Set / Add the generator `PremakeDeps`.
3. Add one (or multiple) "requires" (I used `nlohmann_json/3.11.2` for the example log).
4. Run conan install. I used `conan install . --build missing --output-folder=./Vendor --deploy=full_deploy --settings=build_type=Debug/Release`.
5. conan will fail.
6. No Lua file is generate!

### Logs

```
======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Debug
compiler=msvc
compiler.runtime=dynamic
compiler.runtime_type=Debug
compiler.version=193
os=Windows

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows


======== Computing dependency graph ========
Graph root
    conanfile.py: ...\conanfile.py
Requirements
    nlohmann_json/3.11.2#a35423bb6e1eb8f931423557e282c7ed - Cache

======== Computing necessary packages ========
Requirements
    nlohmann_json/3.11.2#a35423bb6e1eb8f931423557e282c7ed:da39a3ee5e6b4b0d3255bfef95601890afd80709#2d1a5b1f5d673e1dab536bed20ce000b - Cache

======== Installing packages ========
nlohmann_json/3.11.2: Already installed!

======== Finalizing install (deploy, generators) ========
conanfile.py: Conan built-in full deployer to ...\Vendor
conanfile.py: Writing generators to ...\Vendor
conanfile.py: Generator 'PremakeDeps' calling 'generate()'
conanfile.py: ERROR: Traceback (most recent call last):
  File "C:\Program Files\Python311\Lib\site-packages\conans\client\generators\__init__.py", line 70, in write_generators
    generator.generate()
  File "C:\Program Files\Python311\Lib\site-packages\conan\tools\premake\premakedeps.py", line 47, in generate
    generator_files = self.content
                      ^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\conan\tools\premake\premakedeps.py", line 82, in content
    deps = _PremakeTemplate(self._get_cpp_info())
                            ^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\conan\tools\premake\premakedeps.py", line 57, in _get_cpp_info
    dep_cppinfo.aggregate_components()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\conans\model\build_info.py", line 362, in __getattr__
    return getattr(self.components[None], attr)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: '_Component' object has no attribute 'aggregate_components'

ERROR: Error in generator 'PremakeDeps': '_Component' object has no attribute 'aggregate_components'
```

### PR DESCRIPTION
**Pull request by Ohjurot**
**Summary**
Changelog: Feature / Bugfix: Adding support for premake5 (PremakeDeps) fixes #13275 
Docs: https://github.com/conan-io/docs/pull/XXXX

**Features**
- Implements lua file generation for premake consumption (New feature for 2.0)
- Supports multi build with changing architecture and configuration (New for the premake tool)
- Generates a lua file with variables (includedirs, libdirs, ...) per configuration (Based on existing implementation, adapted for multi build and new architecture)
- Generates a lua file with functions (Just build/compiler setup, linker setup and both) for automatic premake project setup per package and profile/config (Breaks compatibility with conan1 but is way more flexible)
- Generates a lua file with functions (see previous) for automatic premake project setup per package
- Generates a global `conandeps.premake5.lua` file with functions (see previous) for automatic premake project setup. 
- Has fallback capabilities to `release` when no fitting conan configuration is available

**Status**
WIP-Status: Code works and has been covered with a basic unit test. Documentation is still required.

**Checklist**
- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conan/tools/premake/premakedeps.py [hunk 1]
```diff
@@ -37,74 +109,200 @@ def __init__(self, deps_cpp_info):
...
+        # Extract all dependencies
         host_req = self._conanfile.dependencies.host
         test_req = self._conanfile.dependencies.test
+        build_req = self._conanfile.dependencies.build
+
+        # Process dependencies and accumulate globally required data
+        pkg_files = []
+        dep_names = []
+        for require, dep in list(host_req.items()) + list(test_req.items()) + list(build_req.items()):
+            dep_name = require.ref.name
```
Comment (memsharded) [inline conan/tools/premake/premakedeps.py:209]: This ``for`` internals is a bit too long, might benefit from some splitting.

File: conan/tools/premake/premakedeps.py [hunk 2]
```diff
@@ -37,74 +109,200 @@ def __init__(self, deps_cpp_info):
...
+    
+    def _output_lua_file(self, filename, content):
+        self.output_files[filename] = "\n".join(["#!lua", *content])
+
+    def _indent_string(self, string, indent=1):
+        return "\n".join([
+            f"{self.tab * indent}{line}" for line in list(filter(None, string.splitlines()))
+        ])
+
+    def _premake_filtered(self, content, configuration, architecture, indent=0):
```
Comment (memsharded) [inline conan/tools/premake/premakedeps.py:159]: These filters are quite unknown for others (like us) not knowing Premake. A short explanation or a link to somewhere that defines what are those would be nice.

**Review by memsharded — COMMENTED** [review]
I think this looks quite good, clean, with templates, new test, etc, good job.

If we want to move this forward for the next patches releases 2.0.X, this PR should be targeted to the ``release/2.0`` branch. As long as we don't document it, it should be very doable.

At some point we might want to add some functional test that really builds something simple, but that can wait a bit if you are running functional testing on your side.

**Comment by Ohjurot** [comment]
@memsharded Thank you for your first feedback. I will close this PR then and open a new one based on the correct branch and targeted towards `release/2.0`.

**Pull request closed without merging**